### ISSUE TITLE
Asyncio support

### ISSUE THREAD

**Issue by oeway** [pyodide/pyodide#245]
Hi, following a discussion here: https://github.com/iodide-project/pyodide/issues/228 I think asyncio is not working right now. 

Would be great to have that supported to call JavaScript functions with async/await in Python.

**Comment by mdboom** [comment]
That's correct.  The default implementation of `asyncio` in Python uses `select`, which isn't available.  This would probably require writing a main loop based on Javascript primitives instead.

In the meantime, you can already use Promises to do most of the things that require async in Javascript, just without the nice syntax.

**Comment by amirouche** [comment]
@mdboom what do you mean by "use Promises" exactly?

AFAIU, One must write an `asyncio` event loop based on [`requestAnimationFrame`](https://developer.mozilla.org/en-US/docs/Web/API/window/requestAnimationFrame), is that correct?

**Comment by mdboom** [comment]
By "use Promises", I mean that the Web APIs that return promises (e.g. `fetch`) work just fine from Python, so you can still do async things without the nice Python async syntax.

I'm not sure what primitives one would build an event loop on -- I had thought `setTimeout`.  But that's the right idea, I think.

**Comment by oeway** [comment]
@mdboom coming back to this question, you mentioned that `select` is not available. Would it be possible to have `select` in pyodide and then `asyncio` will be supported?

**Comment by mdboom** [comment]
One would have to write a select implementation in Javascript built on Web API primitives.  (This is how all the other system calls are implemented in emscripten).  I haven't spent any time looking into how difficult that would be.  Theoretically, if one could write on that's POSIX compliant (presumably only for local "virtual" sockets since remote ones wouldn't work) then Python's asyncio would work.

**Comment by oeway** [comment]
Thanks for the clarification, giving how different(better) it is to use async/await compare to callback style, I would really like to see this feature being supported in the near future. I may also have a look at what you mentioned above, but I am not familiar with these low level libraries.

**Comment by AKKlostermann** [comment]
My idea was not to implement "select" and such, but rather to write a coroutine-based replacement which uses Javascript promises under the hood. This would not support the asyncio libraries, most of which make no sense in a browser anyway.

Instead it would support things like 'sleep'(via a setTimeout wrapper), fetch, websocket messages and so on. A lot of Web APIs can be converted to promises and that would be the easiest way to go.

**Comment by mdboom** [comment]
@AKKlostermann : Sure.  I'd love to see a proof-of-concept of that if you get around to it.

**Comment by oeway** [comment]
@AKKlostermann @mdboom I just want to mention here that I am using js promise as an workaround. 

This is a working demo on ImJoy.io, you just need to click [here](
https://imjoy.io/#/app?plugin=https://gist.githubusercontent.com/oeway/0654908f0e7f0b74ea40cb4ff561940e/raw/88050192b91f61f83cf18c5a622b0e8612c0a6fe/PromiseTimerDemo.imjoy.html
). You can install that demo into your browser and it will load pyodide automatically, if you click "PromiseTimerDemo" in the plugin menu it will use promise to interact with a `setTimeout` defined in JS ([here](https://github.com/oeway/ImJoy/blob/44f0cc4973402b2c898536bf5576803bbcd9bc3d/web/src/components/Imjoy.vue#L1498)).

And this is a code snippet runs in Pyodide:
```python
    def run(self, my):

        api.alert('starting a 5s timer...')

        def timeout_callback(_):
            api.alert('timeout')

        # api.utils.sleep is a remote procedure call
        api.utils.sleep(5).then(timeout_callback)
```

While this could work, it's really hard to debug the code and add lots of overhead to the code. I would really hope there will be a real `async/await` implementation.

**Comment by akloster** [comment]
@mdboom  I hacked together a working proof of concept:
https://github.com/akloster/aioweb-demo/blob/master/src/main.py

It supports running coroutines, awaiting setTimeout and a very basic fetch implementation.

```python
    async def main():
        await sleep(1000)
        request = await Fetcher().get(url)
        obj = await request.json()
```
The "Fetcher" class would be extended to support other options for the browser Fetch API, and request should be an async context manager.

**Comment by mdboom** [comment]
@akloster: I probably won't have a chance to look into this in depth until next week, but just wanted to say this is really great.  Should make doing networking things from Python much easier.

**Comment by akloster** [comment]
I updated my demo with context managers (similar to aiohttp), some error handling, streaming responses via async generators, interacting with react and "requestAnimationFrame".

My next step will be to build a simple development server with aiohttp, which will host the bundles, and hopefully provide a more intelligent package management, especially for the code under development. It will also host API endpoints to exercise the networking functionality some more.

My intention is to have [not necessarily by creating it myself] a pyodide based development framework, suitable for complex single page applications and especially useful for data-centric workloads. I believe that Python's coroutines are a perfect way to interact with promise-based browser APIs and network requests (REST,GraphQL, GRPC etc). I'm looking to interface pyodide with React, because React tends to "get in the way" the least, compared to other modern component systems. Eventually I would like to attach it to the Django Framework with DRF and channels. It would be really nice to have a django-admin like CRUD system for REST APIs, drawing inspiration from Django Models and Forms. Also PixiJS may be a cool supplement for pyodide.

I'd be happy to talk about my ideas some more over pn or email. I'm interested in what other people are working on regarding Python for webassembly. It's certainly early days, but even in the current state, the Download size and initialization speed of Pyodide isn't prohibitive for serious single page applications anymore.

**Comment by AKKlostermann** [comment]
Did anything happen here?

Is there any coordinated effort to push Pyodide forward? If so, by whom and into what directions?

**Comment by rmorshea** [comment]
@AKKlostermann I talked with @mdboom from Mozilla several weeks ago about resolving this issue, but it seems like the folks there don't quite have the cycles to address it at the moment.

I'd like to help review @akloster's work (assuming it's still relevant), but I'd be doing it in my free time, so it's not clear when I'd get to that.

If you're able to contribute though, that'd be great.

**Comment by akloster** [comment]
@rmorshea I'm sorry, @AKKlostermann is a second account of mine ;-)

The demo thing I wrote is really not much more than a small wrapper to integrate Python coroutines with Javascript promises. And probably needs more work to arrive at something useful in a broader sense.

"Blazor" for C# and various Rust "frameworks" are implementing their own virtual DOM and component systems, but I think it would be better for "Python in Webassembly" to leave that to React or a similar library. The Python code would then be tasked with state management, replacing components like Redux or Mobx in a frontend stack. Python's rich datamodel and the concurrency primitives have a lot to offer there. Double that with client-side numerical processing.

To mitigate the problems with download sizes, I think it would be best to have something like a "PyPi as CDN". When multiple websites/applications load packages from the same site, the Browser can cache (and precompile) them more easily. Additionally, the applications could come with flexible version requirements (akin to pip's requirement syntax) such that users don't necessarily need to download every point release of every upstream library. The Service Worker API allows websites to observe and control what is cached and/or prefetched, and that would be necessary to skip an unnecessary update.

**Comment by rmorshea** [comment]
@akloster I agree that Python should leave the VDOM work to React or other libraries - that's actually what I've done with [iDOM](https://github.com/idom) (the project I linked in the issue above) which acts similarly to the state management frameworks you mentioned above.

As far as mitigating download sizes, I'm fairly new to WebAssembly, Pyodide, and Iodide so I'm probably not the best person to advise on that.

**Comment by rmorshea** [comment]
@akloster I'd like to try out your demo some time soon. Do I need to build Pyodide from source in order to create [this pyodide directory](https://github.com/akloster/aioweb-demo/tree/master/src/pyodide)?

**Comment by akloster** [comment]
I think I did that, but I was using their docker configuration. It's been a few weeks that I've worked on it.

**Comment by skelsec** [comment]
Any updates on this topic? I'd like to see this happen

**Comment by oeway** [comment]
I started to explore the possibility of making a custom asyncio event loop and made some progress.

I adapted an [event simulator](https://gist.github.com/damonjw/35aac361ca5d313ee9bf79e00261f4ea) implemented by @damonjw, and now I got this one which runs in Pyodide:  https://alpha.iodide.io/notebooks/6670/ ([Gist](https://gist.github.com/oeway/9a2d5ca5b5920c2929cb05c9ff78e43d))

With this EventSimulator I can use async/await to call a function in Javascript.

For example, in Javascript we have a `asyncFetch` function that returns a promise:
```js
async function asyncFetch(url){
  const response = await fetch(url)
  const data = await response.json()
  return data
}

```

In Python(Pyodide), we can wrap the promise and await the wrapped object:
```python
from js import asyncFetch

def wrap_promise(promise):
    loop = asyncio.get_event_loop()
    fut = loop.create_future()
    promise.then(fut.set_result).catch(fut.set_exception)
    return fut

async def fetchTask():
    print('fetching data...')
    promise = asyncFetch('https://raw.githubusercontent.com/imjoy-team/ImJoy/master/web/package.json')
    result = await wrap_promise(promise)
    print('finished: ',dict(result))

loop = EventSimulator() # implementation here https://gist.github.com/oeway/9a2d5ca5b5920c2929cb05c9ff78e43d
asyncio.set_event_loop(loop)

asyncio.ensure_future(fetchTask())
loop.run_forever()
```
Declaimer: this event loop is a prototype, some features does not behave like the default event loop, there are issues, for example, loop.run_forever() won't block the execution, `asyncio.sleep` is not working properly.

I currently need to wrap the `promise` to make it awaitable, a better implementation is perhaps to add a `__await__` method to the jsProxy object of promise.

@mdboom @akloster  What do you think?

**Comment by hoodmane** [comment]
> a better implementation is perhaps to add a __await__ method to the jsProxy object of promise.

I'm going to try to do this. Mostly it seems to be an issue of copying a bunch of relevant stuff out of [asynciomodule.c](https://github.com/python/cpython/blob/cda99b4022daa08ac74b0420e9903cce883d91c6/Modules/_asynciomodule.c). The coroutine `wrap_promise` is extremely simple so it unwinds into a very simple state machine.

**Comment by oeway** [comment]
UPDATE: The custom event loop fails to run in the new pyodide with python 3.8, we need to explicitly call `asyncio._set_running_loop()`, I have fixed it  in the iodide notebook:  https://alpha.iodide.io/notebooks/6670/  and [Gist](https://gist.github.com/oeway/9a2d5ca5b5920c2929cb05c9ff78e43d).

**Comment by rth** [comment]
Resolved in #1158  https://github.com/pyodide/pyodide/pull/880. We can open follow up issues for more specific points.

**Comment by dberardo-com** [comment]
is it possible to run a blocking function like: loop.run_until_complete(sleep_task())

wait for the results and then return main ?

so basically i need to use fetch() imported from the "js" object and it has only an async API. but i need to use it in an event loop that already exists, so how can i block and wait on the function, as if it was "request.post()" ? for example ?

P.S. i am running pyodide on the backend, not on a browser

**Comment by rmorshea** [comment]
@dberardo-com so your problem basically looks like this?

```python
def some_sync_function():
    # somehow run js.fetch
    response = block_until_complete(js.fetch(...))
    ...

async main():
    some_sync_function()
    
asyncio.run(main())
```

If so, I'm not sure if there are any great solutions being making `some_sync_function` into an async one.

Outside of Pyodide the typical approach to this problem would be to create a thread to run the coroutine, but Pyodide [doesn't support that](https://github.com/pyodide/pyodide/issues/237) AFAIK. There are hacks like [`nest-asyncio`](https://github.com/pyodide/pyodide/issues/237), but this could also have problems since it monkeypatches `asyncio` so you can call `asyncio.run` even if an event loop is already running.

**Comment by dberardo-com** [comment]
after lot of struggling, i have figured out that pyodide is not really meant for backend usage ... unfortunately.

due to the lack of this fundamental features, such as:
* possibility to interact with system executable
* possibility to communicate to external TCP / HTTP services

i had to switch to native python for the implementation i am making 

i hope this changes in future, when folks realize that pyodide can allow usage of python in nodejs apps without the need of installing it.

but for now, thats maybe too early .. thanks for your support, and YES, this is exactly what my problem looks like at the moment (although this is only to face the issue of being not able to use "requests" and having to use "js.fetch"
